_posts/2018/2018-12-17-always-force-booleans-to-be-true-or-false.md

```diff
@@ -0,0 +1,72 @@
+---
+title: Always Force Booleans to be True or False
+description: “Well duh,” you say. “Think about nil,” I say.'
+layout: article
+category: ruby
+image:
+  base: '2018/always-force-booleans-to-be-true-or-false'
+  alt: 'Lovehearts, be true'
+  source: 'https://unsplash.com/photos/ohqX8M_z85E'
+  credit: 'Obi Onyeador'
+---
+
+Migrations, in common with much of Rails, are ‘[sharp knives](https://rubyonrails.org/doctrine/#provide-sharp-knives)’ — easy to use but potentially dangerous and you can cut yourself.
+
+One way a migration can ‘cut’ you is by letting you make quick decisions about your database structure without considering the consequences. In this case we'll examine the case of `NULL` in a boolen field your SQL (converted to `nil` when you use it in Ruby).
+
+
+## Instead of…
+
+...allowing boolean values to be `NULL` in your database.
+
+```ruby
+class AddRecommendedIndexes < ActiveRecord::Migration[5.1]
+  def change
+    add_column :characters, :satisfied, :boolean
+  end
+end
+```
+
+
+## Always…
+
+...use constraints in your database that prevent values that should be only `true` or `false` from being `NULL`.
+
+```ruby
+class AddRecommendedIndexes < ActiveRecord::Migration[5.1]
+  def change
+    add_column :characters, :satisfied, :boolean, default: false, null: false
+  end
+end
+```
+
+
+## But why?
+
+This is a known issue called the tri-state problem. Ideally you don't want a boolean (which by definition should be `true` or `false`) to ever be `NULL` (or `nil` when it is converted from SQL to Ruby).
+
+Although `nil` behaves a lot like `false`, in Ruby they are not quite the same. Having a third state that you can think of as “a bit like `false`” really complicates the logic of your application.
+
+Having `NULL` values in your database means you have to consider how they should behave and also complicates your scopes.
```

Comment:
By this point, It's like you've spend three paragraphs saying: this is not recommended/it's really bad/confusing, etc., without getting to why.

I think you can condense them and get more quickly to: you have to handle them since in Ruby `nil` and `false` are not quite the same and scopes get complicated too, etc.